src/ts/VMhandlers.ts

```diff
@@ -0,0 +1,380 @@
+};
+
+export async function openVM(vm: iVM): Promise<void> {
+	// If there's an active VM it must be closed before opening another
+	if (VM !== null) return;
+	expectedClose = false;
+	// Set hash
+    //@ts-ignore
+	location.hash = vm.id;
+	// Create the client
+    //@ts-ignore
+	VM = new CollabVMClient(vm.url);
+
+	// Register event listeners
+	VM!.on('chat', (username, message) => chatMessage(username, message));
+	VM!.on('adduser', (user) => addUser(user));
+	VM!.on('flag', () => flag());
+	VM!.on('remuser', (user) => remUser(user));
+	VM!.on('rename', (oldname, newname, selfrename) => userRenamed(oldname, newname, selfrename));
+
+	VM!.on('renamestatus', (status) => {
+		// TODO: i18n these
+		switch (status) {
+			case 'taken':
+				alert(TheI18n.GetString(I18nStringKey.kError_UsernameTaken));
+				break;
+			case 'invalid':
+				alert(TheI18n.GetString(I18nStringKey.kError_UsernameInvalid));
+				break;
+			case 'blacklisted':
+				alert(TheI18n.GetString(I18nStringKey.kError_UsernameBlacklisted));
+				break;
+		}
+	});
+
+	VM!.on('turn', (status) => turnUpdate(status));
+	VM!.on('vote', (status: VoteStatus) => voteUpdate(status));
+	VM!.on('voteend', () => voteEnd());
+	VM!.on('votecd', (voteCooldown) => window.alert(TheI18n.GetString(I18nStringKey.kVM_VoteCooldownTimer, voteCooldown)));
+	VM!.on('login', (rank: Rank, perms: Permissions) => onLogin(rank, perms));
+
+	VM!.on('close', () => {
+		if (!expectedClose) alert(TheI18n.GetString(I18nStringKey.kError_UnexpectedDisconnection));
+		closeVM();
+	});
+
+	// auth
+	VM!.on('auth', async server => {
+		elements.changeUsernameBtn.style.display = "none";
+		if (Config.Auth.Enabled && Config.Auth.APIEndpoint === server && auth!.account) {
+			VM!.loginAccount(auth!.account.sessionToken);
+		} else if (!Config.Auth.Enabled || Config.Auth.APIEndpoint !== server) {
+			//@ts-ignore
+			auth = new AuthManager(server);
+			await renderAuth();
+		}
+	});
+
+	// Wait for the client to open
+	await VM!.WaitForOpen();
+
+	// Connect to node
+    //@ts-ignore
+	chatMessage('', `<b>${vm.id}</b><hr>`);
+	let username = Config.Auth.Enabled ? (auth!.account?.username ?? null) : localStorage.getItem('username');
+    //@ts-ignore
+	let connected = await VM.connect(vm.id, username);
+    //@ts-ignore
+	elements.adminInputVMID.value = vm.id;
+    //@ts-ignore
+	w.VMName = vm.id;
+	if (!connected) {
+		closeVM();
+		throw new Error('Failed to connect to node');
+	}
+	// Set the title
+    //@ts-ignore
+	document.title = Format('{0} - {1}', vm.id, TheI18n.GetString(I18nStringKey.kGeneric_CollabVM));
+	// Append canvas
+	elements.vmDisplay.appendChild(VM!.canvas);
+	// Switch to the VM view
+	elements.vmlist.style.display = 'none';
+	elements.vmview.style.display = 'block';
+	return;
+}
+
+export function closeVM() {
+	if (VM === null) return;
+	expectedClose = true;
+	// Close the VM
+	VM.close();
+	VM = null;
+	document.title = TheI18n.GetString(I18nStringKey.kGeneric_CollabVM);
```

Comment:
kGeneric_CollabVM is undefined. This will cause an invariant exception on the webapp.